Add functional support for Docker sub commands on Solaris

amitkris opened the pull request:
This PR will add support for a subset docker sub commands on Solaris.
This PR looks to add support for the following subcommands and explicitly disallow the rest on Solaris.

Supported subcommands:
   attach, build, commit, cp,  create, diff, events ,export, history, images,  import, info, inspect, kill , load, login, logout, logs , network(_null and bridge driver only_), port, ps,  pull, push, rename, restart  , rm, rmi, run, save, search, start, stop, tag, top,  version, wait 

Unsupported commands:
    exec, pause, stats, service, unpause, update, volume

TODO:
- [x] Merge and vendor containerd support [docker/containerd#261](https://github.com/docker/containerd/pull/261)
- [x] Merge and vendor libnetwork support [docker/libnetwork#1225](https://github.com/docker/libnetwork/pull/1225)
- [x] Enable tests (unit tests done, integration tests will be a follow on effort)
- [x] Vendor sys/unix in the Engine for libcontainerd/process_unix.go
- [x] Merge and vendor support for github.com/google/certificate-transparency/go/x509/ https://github.com/google/certificate-transparency/pull/1321

Signed-off-by: Amit Krishnan krish.amit@gmail.com

amitkris committed:
Add funtional support for Docker sub commands on Solaris

Signed-off-by: Amit Krishnan <krish.amit@gmail.com>

justincormack commented:
The CI is complaining about some Go linting issues, might be good to clean up.

thaJeztah commented:
```
08:18:29 # github.com/docker/docker/libcontainerd
08:18:29 libcontainerd/client_linux.go:4: imported and not used: "encoding/json"
08:18:29 libcontainerd/client_linux.go:7: imported and not used: "path/filepath"
08:18:29 libcontainerd/client_linux.go:9: imported and not used: "sync"
08:18:29 libcontainerd/client_linux.go:14: imported and not used: "github.com/docker/docker/pkg/idtools"
08:18:53 Build step 'Execute shell' marked build as failure
```

thaJeztah commented:
ping @amitkris can you rebase? 1.12 is released, so we can begin moving this forward

amitkris commented:
@thaJeztah thats great! thanks for the update. will ping here when done.

amitkris commented:
ping @thaJeztah : rebased changes, its ready to be reviewed.

jhowardmsft commented on daemon/config_windows.go:
```diff
@@ -7,6 +7,12 @@ import (
+const (
+	// stockRuntimeName is the reserved name/alias used to represent the
+	// OCI runtime being shipped with the docker daemon package.
+	stockRuntimeName = "runc"
```
Hmmm. There is no `runc` on Windows....

amitkris replied:
@jhowardmsft : this change was introduced because the stock runtime was being initialized to runc for all platforms and we wanted it to be runz for Solaris. Due to changes that have come about since then this is no longer required, and the logic can go back to what it was.
I'll revert this change.

vdemeester commented on container/memory_store.go:
```diff
@@ -25,8 +28,13 @@ func (c *memoryStore) Add(id string, cont *Container) {
 	c.RLock()
-	res := c.s[id]
+	if runtime.GOOS == "solaris" {
+		res = c.s[id[:12]]
```
This feels a little weird… on solaris, container are stored by *only* the first 12 digits ?

amitkris replied:
I'll remove this. This is a remnant from a time when the runtime on Solaris couldn't handle the full 256 SHA in the container name. That has been fixed. Its been reverted in most places. I missed these two. Thank you for pointing it out!

vdemeester commented on daemon/commit.go:
```diff
@@ -126,8 +126,8 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
 	}
 
 	// It is not possible to commit a running container on Windows
```
This comment should probably be updated.

amitkris replied:
yup

vdemeester commented on daemon/commit.go:
```diff
@@ -126,8 +126,8 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
 	// It is not possible to commit a running container on Windows
-	if runtime.GOOS == "windows" && container.IsRunning() {
-		return "", fmt.Errorf("Windows does not support commit of a running container")
+	if runtime.GOOS == "windows" || runtime.GOOS == "solaris" && container.IsRunning() {
```
And we might want to externalize this condition into a method.

vdemeester commented on daemon/config_solaris.go:
```diff
@@ -17,13 +28,23 @@ type Config struct {
 
 	// Fields below here are platform specific.
-	ExecRoot string `json:"exec-root,omitempty"`
+	ExecRoot       string                   `json:"exec-root,omitempty"`
```
Hum this is more a *note for later* than anything, but these are common with othe `unixes`, we might want to group them :angel:

vdemeester commented on daemon/config_solaris.go:
```diff
@@ -17,13 +28,23 @@ type Config struct {
 type bridgeConfig struct {
 	commonBridgeConfig
+
+	// Fields below here are platform specific.
```
Same here

amitkris replied:
@vdemeester : this has been factored out into config_common_unix.go.

vdemeester commented on daemon/config_solaris.go:
```diff
@@ -32,14 +53,56 @@ func (config *Config) InstallFlags(flags *pflag.FlagSet) {
 
+	config.Runtimes = make(map[string]types.Runtime)
+
+	flags.StringVarP(&config.SocketGroup, "group", "G", "docker", "Group for the unix socket")
```
here too (something like `config.InstallUnixCommonFlags(…)`

vdemeester commented on daemon/container.go:
```diff
@@ -50,7 +51,13 @@ func (daemon *Daemon) GetContainer(prefixOrName string) (*container.Container, e
-	return daemon.containers.Get(containerID), nil
+	var cID string
+	if runtime.GOOS == "solaris" {
+		cID = containerID[:12]
```
same as before, those feel weird :sweat:

amitkris replied:
this will go away :)

vdemeester commented on daemon/container_operations_solaris.go:
```diff
@@ -15,12 +19,40 @@ func (daemon *Daemon) setupLinkedContainers(container *container.Container) ([]s
 // getSize returns real size & virtual size
 func (daemon *Daemon) getSize(container *container.Container) (int64, int64) {
-	return 0, 0
+	var (
```
Same as the *note for later*, this is duplicated from `container_operations_unix.go`, we should find a way to not duplicate code :angel:

amitkris replied:
Agreed. It'll be easier to track once this PR gets in. Every time I rebase the PR there are some scenarios such as this one. I do my due diligence but sometimes.. :)

vdemeester commented on daemon/container_operations_solaris.go:
```diff
@@ -45,5 +107,15 @@ func detachMounted(path string) error {
 
 func isLinkable(child *container.Container) bool {
+	// A container is linkable only if it belongs to the default network
+	_, ok := child.NetworkSettings.Networks["bridge"]
```
`_, ok := child.NetworkSettings.Networks[runconfig.DefaultDaemonNetworkMode().NetworkName()]` feels safer (and is what's use in `container_operations_unix.go`).

vdemeester commented on daemon/daemon.go:
```diff
@@ -593,7 +593,7 @@ func NewDaemon(config *Config, registryService registry.Service, containerdRemot
 	// Check if Devices cgroup is mounted, it is hard requirement for container security,
 	// on Linux.
-	if runtime.GOOS == "linux" && !sysInfo.CgroupDevicesEnabled {
+	if runtime.GOOS != "windows" && runtime.GOOS != "solaris" && !sysInfo.CgroupDevicesEnabled {
```
I'm not sure about this change. `solaris` is not seen as `linux` right, this condition shouldn't need any change, right ?

amitkris replied:
hmm..let me test it out.

vdemeester commented on daemon/daemon_solaris.go:
```diff
@@ -63,13 +113,199 @@ func (daemon *Daemon) getCgroupDriver() string {
+	}
+
+	if hostConfig.ShmSize != 0 {
+		shmSize := container.DefaultSHMSize
```
absolutely a nit, but `hostConfig.ShmSize = container.DefaultSHMSize`

amitkris replied:
noted

vdemeester commented on daemon/daemon_solaris.go:
```diff
@@ -79,6 +315,16 @@ func (daemon *Daemon) platformReload(config *Config, attributes *map[string]stri
+		config.Runtimes = make(map[string]types.Runtime)
+	}
+	stockRuntimeOpts := []string{}
+	config.Runtimes[stockRuntimeName] = types.Runtime{Path: "runc", Args: stockRuntimeOpts}
```
Should we use `DefaultRuntimeBinary` instead of plain "runc" here ?

amitkris replied:
absolutely. my bad. will change this.

vdemeester commented on daemon/stats.go:
```diff
@@ -19,8 +18,8 @@ import (
 func (daemon *Daemon) ContainerStats(ctx context.Context, prefixOrName string, config *backend.ContainerStatsConfig) error {
-	if runtime.GOOS == "windows" {
-		return errors.New("Windows does not support stats")
+	if runtime.GOOS == "windows" || runtime.GOOS == "solaris" {
```
I think windows supports stats now :angel:

amitkris replied:
thank you for pointing that out!

vdemeester commented on libcontainerd/client_unix.go:
```diff
@@ -0,0 +1,147 @@
+// +build solaris linux
```
Not sure if these build tags are needed at all. It should work the same without it (and keeping *supporting* freebsd and other unix somehow)

vdemeester commented on libcontainerd/container_unix.go:
```diff
@@ -1,3 +1,5 @@
+// +build linux solaris
```
Same here

vdemeester commented on libcontainerd/pausemonitor_unix.go:
```diff
@@ -1,3 +1,5 @@
+// +build linux solaris
```
And here

vdemeester replied:
and below in this package.

amitkris replied:
will revisit the build tags. thanks for your comments @vdemeester 
sorry about the delay in replying to them. Was on vacation earlier this week.
Will update the changeset shortly.

amitkris commented on container/container_unix.go:
```diff
@@ -320,6 +320,7 @@ func (container *Container) UpdateContainer(hostConfig *containertypes.HostConfi
 }
 
 func detachMounted(path string) error {
+	// TODO: replace call to syscall.Unmount with unix.Unmount after sys/unix has been vendored.
```
not yet. Will file it and tag it to this PR.

amitkris committed:
Uprev golang.org/x/sys to vendor in sys/unix which is required for Solaris support.
This is because the Solaris platform does not support the syscall package.
Note that while Solaris supports all the common functionality in sys/unix there are platform specific features (for eg. MNT_DETACH) which are not implemented for the Solaris platform.

Signed-off-by: Amit Krishnan <krish.amit@gmail.com>

amitkris commented:
@vdemeester : I've updated the changeset and incorporated all your comments.

I added a changeset for the sys/unix changes. Thats causing the vendor CI to fail. Its failing with "directory not empty". Have you seen that issue before?
Also the build tags you mentioned for libcontainerd/*_unix.go is causing the windowsRS1 CI to fail. I tried both with and without the build tags. Is there aynthing additional that needs to be added for that?

thaJeztah commented:
ping @amitkris I see the linked PR's are merged; can you vendor and update the PR? Code freeze for 1.13 is next week

amitkris commented:
@thaJeztah : I was in the process of addressing the action items :)
the libnetwork uprev is already taken care of in #27594 
The sys/unix package is vendored in as an additional changeset in this PR itself.
I've added (and squashed) the unit test changes into the main changeset itself.

thaJeztah commented:
Perfect, thanks @amitkris

amitkris committed:
Vendor in libnetwork/drivers/solaris/bridge

Signed-off-by: Amit Krishnan <krish.amit@gmail.com>

amitkris committed:
Uprev github.com/tonistiigi/fifo/fifo.go

Signed-off-by: Amit Krishnan <krish.amit@gmail.com>

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "solaris-1.12" git@github.com:amitkris/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~4
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

tiborvass commented on daemon/cluster/listen_addr_solaris.go:
```diff
@@ -0,0 +1,23 @@
+	defRouteCmd := "/usr/sbin/ipadm show-addr -p -o addr " +
+		"`/usr/sbin/route get default | /usr/bin/grep interface | " +
+		"/usr/bin/awk '{print $2}'`"
+	out, err := exec.Command("/usr/bin/bash", "-c", defRouteCmd).Output()
```
Would be great to find a way without shelling out

tiborvass commented on daemon/daemon_solaris.go:
```diff
@@ -27,12 +37,52 @@ const (
+		until we figure out how to map to Trusted Extensions
+		this is being disabled for now on Solaris
+	*/
+	var (
```
please remove dead code

tiborvass commented on registry/auth_test.go:
```diff
@@ -1,3 +1,5 @@
+// +build !solaris
+
```
```
// +build !solaris

// TODO: support solaris

package registry
```

amitkris replied:
done

tiborvass commented on pkg/archive/archive_test.go:
```diff
@@ -68,6 +68,9 @@ func TestIsArchivePathDir(t *testing.T) {
 func TestIsArchivePathInvalidFile(t *testing.T) {
 	cmd := exec.Command("sh", "-c", "dd if=/dev/zero bs=1K count=1 of=/tmp/archive && gzip --stdout /tmp/archive > /tmp/archive.gz")
+	if runtime.GOOS == "solaris" {
+		cmd = exec.Command("sh", "-c", "dd if=/dev/zero bs=1 count=1 of=/tmp/archive && gzip --stdout /tmp/archive > /tmp/archive.gz")
```
factor out `bs=1`/`bs=1K` to remove the `if`

amitkris replied:
done

tiborvass commented on pkg/archive/archive_test.go:
```diff
@@ -81,7 +84,12 @@ func TestIsArchivePathInvalidFile(t *testing.T) {
-	cmd := exec.Command("sh", "-c", "touch /tmp/archivedata && tar -cf /tmp/archive /tmp/archivedata && gzip --stdout /tmp/archive > /tmp/archive.gz")
+	var cmd *exec.Cmd
+	if runtime.GOOS == "solaris" {
+		cmd = exec.Command("sh", "-c", "touch /tmp/archivedata && gtar -cf /tmp/archive /tmp/archivedata && gzip --stdout /tmp/archive > /tmp/archive.gz")
```
factor out `tar`/`gtar` to remove the if

amitkris replied:
done

tiborvass commented on libcontainerd/process_unix.go:
```diff
@@ -1,22 +1,25 @@
 	"os"
 	"path/filepath"
-	"syscall"
+	runt "runtime"
```
please rename to `goruntime`

amitkris replied:
done

tiborvass commented on daemon/daemon_solaris.go:
```diff
@@ -27,12 +37,52 @@ const (
+		err       error
+	)
+
+	for range config.SecurityOpt {
```
```
if len(config.SecurityOpt) > 0 {
    return ...
}
```

amitkris replied:
done

tiborvass commented:
Overall looks good. Just need to address small nits and vendor the new google certificate dependency.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "solaris-1.12" git@github.com:amitkris/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~5
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

amitkris commented:
@tiborvass: thanks for the comments. addressed all of your and @vdemeester's comments and rebased over the newly vendored certificate-transparency.

amitkris commented:
@jhowardmsft, @thaJeztah  : ping
Can you help me figure out the windowsRS1 CI issues?

Its failing at this step:
21:50:12 Step 1/6 : FROM windowsservercore
21:50:12  ---> 01569f895bac
21:50:12 Step 2/6 : RUN mkdir C:\tmp && mkdir C:\busybox
21:50:12  ---> Running in fed9ae364537
21:50:14 windows does not support commit of a running container

amitkris commented:
@thaJeztah : fixed the issue. CI is now green :)

vdemeester commented:
LGTM 🐸

tiborvass commented:
@amitkris question: what happens if a container was packaged when running on an older solaris kernel and is now being run on a newer solaris kernel? The container has the old libc, and I was told ABIs are not stable (one has to use the corresponding libc)

amitkris commented:
@tiborvass : In Solaris we have what are called boot environments (B.E's). Every time you update the kernel you get a new B.E. A container that is baked on a certain B.E. stays with that B.E. We manage it such that when you switch between B.E.'s only the containers that are compatible with that B.E. will show up.

justincormack commented on container/container_notsolaris.go:
```diff
@@ -0,0 +1,11 @@
+// +build linux freebsd
```
This only actually builds on Linux, FreeBSD does not have `unix.MNT_DETACH`, so looks like this should be `container_linux` and FreeBSD and Solaris can share the other file.

If you don't want to change this, feel free to leave it and I will try to tidy up the mess that is our bitrotted FreeBSD support later...

amitkris replied:
I've changed the files to container_linux.go and container_notlinux.go

justincormack commented on daemon/cluster/listen_addr.go:
```diff
@@ -270,6 +279,39 @@ func listSystemIPs() []net.IP {
 	return systemAddrs
 }
 
+func listSolarisSystemIPs() []net.IP {
```
this should be split out into a conditionally compiled file ideally

justincormack replied:
no fine to keep it temporarily

justincormack commented on daemon/commit.go:
```diff
@@ -126,9 +126,9 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
-	if runtime.GOOS == "windows" && container.IsRunning() {
-		return "", fmt.Errorf("Windows does not support commit of a running container")
+	// It is not possible to commit a running container on Windows and on Solaris.
+	if (runtime.GOOS == "windows" || runtime.GOOS == "solaris") && container.IsRunning() {
```
It would be nicer to refactor this as a set of daemon feature flags eg `daemon.feature.commitRunning = false` and test these rather than OS conditionals, then collect them all together in a per OS file.

amitkris replied:
Honestly I think that would be overkill for one feature.. A multi-platform check such as this is done only in this file (all others are in test files and not feature specific). If such checks show up for multiple features in the future I'd be happy to implement that. For now I think its most readable this way.

justincormack replied:
This is not the only feature, there are more, but happy if we fix later.

As a matter of principle, we should not have references to `runtime` anywhere. These are static features.

justincormack commented on daemon/daemon_solaris.go:
```diff
@@ -67,13 +118,198 @@ func (daemon *Daemon) getCgroupDriver() string {
 // hostconfig and config structures.
 func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.HostConfig, config *containertypes.Config, update bool) ([]string, error) {
 	warnings := []string{}
+	sysInfo := sysinfo.New(true)
```
Not sure about ignoring these values, I think it might be better to error and fail.

amitkris replied:
Thats a good point. Some of the features here have been stubbed pending certain minor changes in Solaris OS code which are being worked upon. Once we have a fully threshed out list of what we're looking to support going forward we will error out on those unsupported. 
So until then we'd like to keep it this way.

justincormack replied:
ok

justincormack commented on pkg/archive/archive_test.go:
```diff
@@ -67,7 +67,14 @@ func TestIsArchivePathDir(t *testing.T) {
-	cmd := exec.Command("sh", "-c", "dd if=/dev/zero bs=1K count=1 of=/tmp/archive && gzip --stdout /tmp/archive > /tmp/archive.gz")
+	var bsVal string
+	if runtime.GOOS == "solaris" {
+		bsVal = "bs=1"
```
Are you sure? The man page says that would be a 1 byte blocksize! `bs=1024` should work for all platforms.

justincormack commented:
@amitkris hi, have looked through this and added some comments. Generally I would be happy if most/all of them were addressed in followup PRs instead and I think it would be ok to merge as is.

amitkris commented on hack/make/cross:
```diff
@@ -29,15 +29,18 @@ for platform in $DOCKER_CROSSPLATFORMS; do
-			source "${MAKEDIR}/binary-client"
-			source "${MAKEDIR}/binary-daemon"
+		if [ "$GOOS" != "solaris" ]; then
+			# TODO. Solaris fails CI because of CGO calls.
```
Yes. I've updated the comment.

amitkris commented:
@justincormack : Hi! Thank you for the comments :)
I've addressed some of your comments and replied to the rest. I've added an additional changeset with the changes. Please let me know what you think.

thaJeztah commented:
@amitkris looks like it needs a rebase

amitkris commented:
@thaJeztah : done

amitkris commented:
thanks @tiborvass !

thaJeztah commented:
oh, @amitkris is going to squash the last commit before merging, so hold your horses :smile:

amitkris commented:
thanks @justincormack! squashing the topmost changeset now.

thaJeztah commented:
Rebased/Carried in https://github.com/docker/docker/issues/28056

tiborvass commented:
In that case I'm closing this one.
